include/merlin/memory_pool.cuh

```diff
@@ -0,0 +1,600 @@
+ */
+template <class T>
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    delete[] ptr;
+  }
+};
+
+/**
+ * Claim/release buffers in pinned host memory.
+ */
+template <class T>
+struct HostAllocator final : AllocatorBase<T, HostAllocator<T>> {
+  using type = typename AllocatorBase<T, HostAllocator<T>>::type;
+
+  static constexpr const char* name{"HostAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    CUDA_CHECK(cudaMallocHost(&ptr, sizeof(T) * n));
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    CUDA_CHECK(cudaFreeHost(ptr));
+  }
+};
+
+/**
+ * Claim/release buffers in the active CUDA device. Will not test if the correct
+ * device was used, and throw if CUDA runtime API response is negative.
+ */
+template <class T>
+struct DeviceAllocator final : AllocatorBase<T, DeviceAllocator<T>> {
+  using type = typename AllocatorBase<T, DeviceAllocator<T>>::type;
+
+  static constexpr const char* name{"DeviceAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    cudaError_t res;
+    if (ctx) {
+      res = cudaMallocAsync(&ptr, sizeof(T) * n, ctx);
+    } else {
+      res = cudaMalloc(&ptr, sizeof(T) * n);
+    }
+    CUDA_CHECK(res);
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    cudaError_t res;
+    if (ctx) {
+      res = cudaFreeAsync(ptr, ctx);
+    } else {
+      res = cudaFree(ptr);
+    }
+    CUDA_CHECK(res);
+  }
+};
+
+/**
+ * Wrapper around another allocator that prints debug messages.
+ */
+template <class Allocator>
+struct DebugAllocator final
+    : AllocatorBase<typename Allocator::type, DebugAllocator<Allocator>> {
+  using type = typename Allocator::type;
+
+  static constexpr const char* name{"DebugAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    type* ptr = Allocator::alloc(n, ctx);
+    std::cout << Allocator::name << "[type_name = " << typeid(type).name()
+              << "]: " << static_cast<void*>(ptr) << " allocated = " << n
+              << " x " << sizeof(type) << " bytes, ctx = " << ctx << std::endl;
+    return ptr;
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    Allocator::free(ptr, ctx);
+    std::cout << Allocator::name << "[type_name = " << typeid(type).name()
+              << "]: " << static_cast<void*>(ptr) << " freed, ctx = " << ctx
+              << std::endl;
+  }
+};
+
+/**
+ * Helper structure to configure a memory pool.
+ */
+struct MemoryPoolOptions {
```

Comment:
This struct should be exposed to users, so may you should move the declaration to `merloin_hashtable.cuh`? or make the API doc generator can see it.

Reply:
I would like to object in this case.
* **Technical argument**: I would like to keep it away from a `.cuh` file. The entire resource management can compile with GCC.
* **Compile sequence**: The resource manager doesn't depend on on the hashtable. But the hashtable depends on thee resource manager. Thus, `merlin_hashtable.cuh` includes `memory_pool.hpp`, not the other way around.
* **Separation of concerns**: `merlin_hashtable.cuh` focuses on hashtable stuff. The `memory_pool.hpp` just considers resource allocation.

Reply:
With the switch to `cuh` file and NVCC compilation the `technical argument` has become moot. 

Compile sequence is still an issue. `merlin_hashtable` depends on `memory pool` and therefore has to include it. So `memory pool` cannot include `merlin_hashtable`. I suggest we create a header for option structs.